### ISSUE TITLE
Freezing weights

### ISSUE THREAD

**Issue by giamic** [OpenNMT/OpenNMT-py#1418]
I'm training a transformer model on MIDI music files, but results are poor because I only have access to a small dataset pertaining to the style I want to study. To help the model learn something useful, I would like to use a much larger dataset of other styles of music for a pre-training and then fine-tune the results using the small dataset.

I was thinking of freezing the encoder side of the transformer after the pre-training and letting the decoder part free to do the fine-tuning. Is this currently possible? Please note that I don't want to freeze just the embedding parts, but the entire set of attention weights in the encoder.

**Comment by vince62s** [comment]
It is not straight forward to accomplish this.
In the current code base, the whole model est forward then backpropagated. (have a look at the training loop in trainer.py)
There could be a way if you implement a detach_state() function in the encoder in the same way as it is for the decoder (used in the case of bptt). You'll really have to understand the code to accomplish this.
Good luck.

**Comment by eraoul** [comment]
I'm curious about this too. This mostly makes sense.

I think one could just add in a new option flag such as `freeze_encoder` and then if it's `True`, run something like the `detach_state` function in the encoder.

In this case, though, is it better to just use `requires_grad = False` for the encoder (instead of `detach()`) because we don't need to compute the gradients at all? 

I'm relatively new to pytorch -- trying to understand the difference: (e.g., here: https://discuss.pytorch.org/t/detach-no-grad-and-requires-grad/16915 )

**Comment by giamic** [comment]
Hi, I'm following with interest @eraoul 's approach and I think that it makes sense. Another possibility is to save the Transformer model in two separate files, encoder and decoder, then use the encoder part as a normal function and the decoder model in training mode. Provided what I'm saying actually gives the correct result, it is conceptually more convoluted but maybe easier from the implementation point of view? (that's not a rhetoric question - I'm not familiar with pytorch at all)

**Comment by vince62s** [comment]
Everything is possible...it's just coding.
requiers_grad = False is not possible at the moment without disconnecting encoder/decoder as @giamic says indirectly.
the other solution is not straight forward either.
Anyway, closing this for now. Maybe discuss this further on the forum untill someone picks this up for a PR.
thanks.

**Comment by eraoul** [comment]
Moved discussion to forum as requested:

http://forum.opennmt.net/t/transformer-freezing-encoder-while-training-decoder/2723
